src/mesh/rbf.f90

```diff
@@ -0,0 +1,597 @@
+      DO i=1,Side%nCurvedNodes
+        IF ((x(1).LT.xMin).OR.(x(1).GT.xMax).OR.(x(2).LT.yMin).OR.(x(2).GT.yMax)) CYCLE
+        IF (Side%CurvedNode(i)%np%tmp.EQ.1) CYCLE
+        iBP = iBP + 1
+        RefCoordinates(1:3,iBP) = Side%CurvedNode(i)%np%x
+        RBFRHS(iBP,1:3) = 0.
+        Side%CurvedNode(i)%np%tmp = 1
+      END DO
+    ELSE
+      ! Curved surfaces
+      x = Side%CurvedNode(QuadMapInv(0,0))%np%x
+      IF ((x(1).LT.xMin).OR.(x(1).GT.xMax).OR.(x(2).LT.yMin).OR.(x(2).GT.yMax)) EXIT
+      ! Build coordinates of linear side. First, fill an temp array with the corner nodes
+      xCornerSurf(:,QuadMapInvLinear(0,0)) = Side%CurvedNode(QuadMapInv(0,0))%np%x
+      xCornerSurf(:,QuadMapInvLinear(0,1)) = Side%CurvedNode(QuadMapInv(0,N))%np%x
+      xCornerSurf(:,QuadMapInvLinear(1,0)) = Side%CurvedNode(QuadMapInv(N,0))%np%x
+      xCornerSurf(:,QuadMapInvLinear(1,1)) = Side%CurvedNode(QuadMapInv(N,N))%np%x
+      ! Then, apply Vandermonde matrix, e.g. evaluate bi-linear mapping on all curved side nodes
+      xBiLinear(1,:) = MATMUL(Vdm_SurfBiLinear,xCornerSurf(1,:))
+      xBiLinear(2,:) = MATMUL(Vdm_SurfBiLinear,xCornerSurf(2,:))
+      xBiLinear(3,:) = MATMUL(Vdm_SurfBiLinear,xCornerSurf(3,:))
+      DO i=1,Side%nCurvedNodes
+        ! Skip already marked boundary points
+        IF (Side%CurvedNode(i)%np%tmp.EQ.1) CYCLE
+        ! Increase index of boundary points
+        iBP = iBP + 1
+        ! Store reference coordinates, e.g. the bi-linear nodes
+        RefCoordinates(1:3,iBP) = xBiLinear(:,i)
+        ! Calculate displacement, e.g. difference between bi-linear and curved nodes
+        RBFRHS(iBP,1:3) = Side%CurvedNode(i)%np%x - xBiLinear(:,i)
+        ! Mark this side node as done
+        Side%CurvedNode(i)%np%tmp = 1
+      END DO
+    END IF
+    Side=>Side%nextElemSide
+  END DO
+  Elem=>Elem%nextElem
+END DO
+
+! Set the rows of the R.H.S. associated with the linear polynomial to zero
+RBFRHS(nBP+1:nBP+4,:) = 0.
+
+!--------------------------------------------------------------------------------------------------------!
+! Solve the RBF system
+!--------------------------------------------------------------------------------------------------------!
+WRITE(*,*) 'Solving RBF system... '
+! Build the RBF matrix
+! Upper left part: M_i,j = phi(abs(X(i)-X(j)))
+DO j=1,nBP; DO i=1,nBP
+  dist = Distance(RefCoordinates(:,i),RefCoordinates(:,j))
+  RBFMatrix(i,j) = EvaluateRBF(dist,SupportRadius(iRBFBox),RBFType(iRBFBox))
+END DO; END DO ! j,i=1,nBP
+! Upper right part: P_b - row i equal to [1 x_i y_i z_i]
+! Lower left part: transpose of P_b
+DO i=1,nBP
+   RBFMatrix(i,nBP+1) = 1.
+   RBFMatrix(nBP+1,i) = 1.
+   RBFMatrix(i,nBP+2) = RefCoordinates(1,i)
+   RBFMatrix(nBP+2,i) = RefCoordinates(1,i)
+   RBFMatrix(i,nBP+3) = RefCoordinates(2,i)
+   RBFMatrix(nBP+3,i) = RefCoordinates(2,i)
+   RBFMatrix(i,nBP+4) = RefCoordinates(3,i)
+   RBFMatrix(nBP+4,i) = RefCoordinates(3,i)
+END DO ! i=1,nBP
```

Comment:
Maybe combine these two loops? The second one seems to overwrite a lot of the entries set by the first loop.

Reply:
I do not think that it overwrites anything tbh. It's only the last four rows and columns (nRB+1 to nRB+4) which are filled.

Reply:
RESOLVED: You are right, I misread the second index.